mantidproject/mantid
# Fix finalization of algorithms when running processGroups()

Fixes the finalization of algorithms when running `processGroups()`. `Algorithm::execute()` was not fully finalizing algorithms when running on groups (`processGroups()`). Then several algorithms were doing things like this at the end of their redefinition of `processGroups()`: 

```
// Store output workspace in AnalysisDataService
if (!isChild())
  this->store();

setExecuted(true);
notificationCenter().postNotification(
   new FinishedNotification(this, this->isExecuted()));
```

Other algorithms, line `SaveNXTomo` were not setting the executed flag and not sending the `FinishedNotification` so they wouldn't finalize correctly. The changes here make sure that those finalization steps are done by `Algorithm::execute()` as when running the normal `exec()`.

**To test:**


- Review code changes and see if you like the solution used here.
- All tests should pass and algorithms that process groups should still work well.

To see the problem raised by #11361 and #11537 you can run something similar to the [SaveNXTomo doc example](http://docs.mantidproject.org/v3.5.1/algorithms/SaveNXTomo-v1.html):

```
LoadFITS('FITS_small_01.fits', OutputWorkspace='reconstruction')
LoadFITS('FITS_small_02.fits', OutputWorkspace='reconstruction')
SaveNXTomo(InputWorkspaces='reconstruction', Filename='input_rec')
```

If you run this before the changes in this branch, the algorithm runs but it doesn't finish (in the sense of sending a `FinishedNotification`). In the original issue description the progress bar didn't move. That was fixed recently. The current behavior is that the progress bar reaches 100% and freezes there. Then, if you try to close Mantid it will not close but stay "shutting down..." as it waits for SaveNXTomo to finish.

Fixes #11361.
Also fixes #11537

_Does not need to be in the release notes._

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
##### Functional Tests
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
